oauthenticator/google.py

```diff
@@ -314,6 +315,31 @@ async def check_allowed(self, username, auth_model):
         # users should be explicitly allowed via config, otherwise they aren't
         return False
 
+    def _get_credentials(self, user_email_domain):
+        """
+        Returns the stored credentials or fetches and stores new ones.
+
+        Checks if the credentials are valid before returning them. Refreshes
+        if necessary and stores the refreshed credentials.
+        """
+        if not self._credentials or not self._is_token_valid():
+            self._credentials = self._setup_credentials(user_email_domain)
```

Comment:
Credentials seems to be received for a `user_email_domain`, doesn't that makes them different between users? If so, isn't it causing issues to re-use them for any user?

I figure we would at least need to have a dictionary of credentials, with one for each user_email_domain, right?

Reply:
If multiple people login at the same time `_setup_credentials` could be called multiple times in parallel. Is this likely to cause any problems?

Reply:
Ha, this totally flew past me. All our hubs are single domain, so it didn't occur to me. Will fix!